pkg/conversion/converter.go

```diff
@@ -0,0 +1,245 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+)
+
+func NewConverter(
+	crd *apiextensionsv1.CustomResourceDefinition,
+	apiConversion *apisv1alpha1.APIConversion,
+) (*Converter, error) {
+	structuralSchemas := make(map[string]*structuralschema.Structural)
+
+	// Gather all the structural schemas for easy map-based indexing
+	for _, v := range crd.Spec.Versions {
+		if v.Schema == nil {
+			continue
+		}
+
+		internalJSONSchemaProps := &apiextensionsinternal.JSONSchemaProps{}
+		if err := apiextensionsv1.Convert_v1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(v.Schema.OpenAPIV3Schema, internalJSONSchemaProps, nil); err != nil {
+			return nil, fmt.Errorf("failed converting version %s validation to internal version: %w", v.Name, err)
+		}
+
+		structuralSchema, err := structuralschema.NewStructural(internalJSONSchemaProps)
+		if err != nil {
+			return nil, fmt.Errorf("error getting structural schema for version %s: %w", v.Name, err)
+		}
+
+		if err := structuraldefaulting.PruneDefaults(structuralSchema); err != nil {
+			return nil, fmt.Errorf("error pruning defaults for version %s: %w", v.Name, err)
+		}
+
+		structuralSchemas[v.Name] = structuralSchema
+	}
+
+	compiledRules, err := CompileConversions(apiConversion, structuralSchemas)
+	if err != nil {
+		return nil, fmt.Errorf("error compiling conversion rules: %w", err)
+	}
+
+	return &Converter{
+		crd:               crd,
+		compiledRules:     compiledRules,
+		structuralSchemas: structuralSchemas,
+		apiConversion:     apiConversion,
+	}, nil
+}
+
+type Converter struct {
+	crd               *apiextensionsv1.CustomResourceDefinition
+	compiledRules     map[string][]*CompiledRule
+	structuralSchemas map[string]*structuralschema.Structural
+	apiConversion     *apisv1alpha1.APIConversion
+}
+
+func (c *Converter) Convert(list *unstructured.UnstructuredList, targetGV schema.GroupVersion) (*unstructured.UnstructuredList, error) {
+	ctx := context.TODO()
+
+	convertedList := &unstructured.UnstructuredList{}
+	for i := range list.Items {
+		original := &list.Items[i]
+
+		converted, err := c.convert(ctx, original, targetGV)
+		if err != nil {
+			return nil, err
+		}
+
+		convertedList.Items = append(convertedList.Items, *converted)
+	}
+
+	return convertedList, nil
+}
+
+func (c *Converter) convert(ctx context.Context, original *unstructured.Unstructured, targetGV schema.GroupVersion) (*unstructured.Unstructured, error) {
+	converted := original.DeepCopy()
+	converted.SetAPIVersion(targetGV.String())
+
+	originalVersion := original.GetObjectKind().GroupVersionKind().Version
+	compiledRules := c.compiledRules[originalVersion]
+	for _, rule := range compiledRules {
+		value, err := evaluateRule(ctx, original, rule)
+		if err != nil {
+			return nil, fmt.Errorf("error converting: %w", err)
+		}
+
+		toPath := rule.ToPath
```

Comment:
Do we need to keep the string-formatted value in `rule`? Why not store the array?

Reply:
You mean during rule compilation?